docs/source/cpp/csv.rst

```diff
@@ -0,0 +1,142 @@
+.. KIND, either express or implied.  See the License for the
+.. specific language governing permissions and limitations
+.. under the License.
+
+.. default-domain:: cpp
+.. highlight:: cpp
+
+.. cpp:namespace:: arrow::csv
+
+=================
+Reading CSV files
+=================
+
+Arrow provides a fast CSV reader allowing ingestion of external data
+as Arrow tables.
+
+Basic usage
+===========
+
+A CSV file is read from a :class:`~arrow::io::InputStream`.
+
+.. code-block:: cpp
+
+   #include "arrow/csv/api.h"
+
+   {
+      // ...
+      arrow::Status st;
+      arrow::MemoryPool* pool = default_memory_pool();
+      std::shared_ptr<arrow::io::InputStream> input = ...;
+
+      auto read_options = arrow::csv::ReadOptions::Defaults();
+      auto parse_options = arrow::csv::ParseOptions::Defaults();
+      auto convert_options = arrow::csv::ConvertOptions::Defaults();
+
+      // Instantiate TableReader from input stream and options
+      std::shared_ptr<arrow::csv::TableReader> reader;
+      st = arrow::csv::TableReader::Make(pool, input, read_options,
+                                         parse_options, convert_options,
+                                         &reader);
+      if (!st.ok()) {
+         // Handle TableReader instantiation error...
+      }
+
+      std::shared_ptr<arrow::Table> table;
+      // Read table from CSV file
+      st = reader->Read(&table);
+      if (!st.ok()) {
+         // Handle CSV read error
+         // (for example a CSV syntax error or failed type conversion)
+      }
+   }
+
+Column names
+============
+
+There are three possible ways to infer column names from the CSV file:
+
+* By default, the column names are read from the first row in the CSV file
+* If :member:`ReadOptions::column_names` is set, it forces the column
+  names in the table to these values (the first row in the CSV file is
+  read as data)
+* If :member:`ReadOptions::autogenerate_column_names` is true, column names
+  will be autogenerated with the pattern "f0", "f1"... (the first row in the
+  CSV file is read as data)
+
+Column selection
+================
+
+By default, Arrow reads all columns in the CSV file.  You can narrow the
+selection of columns with the :member:`ConvertOptions::include_columns`
+option.  If some columns in :member:`ConvertOptions::include_columns`
+are missing from the CSV file, an error will be emitted unless
+:member:`ConvertOptions::include_missing_columns` is true, in which case
+the missing columns are assumed to contain all-null values.
+
+Interaction with column names
+-----------------------------
+
+If both :member:`ReadOptions::column_names` and :member:`ConvertOptions::include_columns`,
+the :member:`ReadOptions::column_names` are assumed to map to CSV columns,
+and :member:`ConvertOptions::include_columns` is a subset of those column
+names that will part of the Arrow Table.
+
+Data types
+==========
+
+It is possible to convert a CSV column into the following Arrow data types:
```

Comment:
Is it easier to describe/list the data types that can't be read from a CSV?

Reply:
I don't think so. I find it more useful to list existing functionality, not missing functionality :-)
Also, unsupported data types are more likely to get outdated when adding new Arrow data types without adding CSV support for them. :-)

Reply:
Ok, so `float32` but no `float16`? `timestamp` but no `date*` or `time*`? No dictionary? I could see not being able to infer them, but I'm surprised I can't specify them in `column_types`.

Reply:
Well, there's no converted written for them.